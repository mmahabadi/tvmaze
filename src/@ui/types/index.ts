export type SwiperOptions = {
  centeredSlides: boolean
  slidesPerView: number
  spaceBetween: number
  navigation: boolean
  loop: boolean
  scrollbar: { draggable: boolean }
  speed: number
  lazy: boolean
  autoplay: {
    delay: number
    disableOnInteraction: boolean
  }
}
