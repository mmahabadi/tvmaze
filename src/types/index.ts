export type AppSettings = {
  itemsInEachGenreSection: number
  itemsInEachPage: number
  itemsToShowInHomePageSlider: number
  yearLimitToConsiderAsRecentToShowInHomePageSlider: number
}

export type Rating = {
  average: number
}

export type Schedule = {
  time: string
  days: string[]
}

export type Image = {
  medium: string
  original: string
}

export type ShowByGenre = {
  genre: string
  movies: Movie[]
  loading: boolean
  error: null | Error
}

export type Movie = {
  id: number
  name: string
  type: string
  language: string
  genres: string[]
  status: string
  ended: string
  officialSite: string
  schedule: Schedule
  rating: Rating
  image: Image
  summary: string
  _embedded: Embedded
}

export type Embedded = {
  cast: Cast[]
}

export type Cast = {
  person: Person
  character: Character
}

export type Person = {
  id: number
  url: string
  name: string
  country: Country
  birthday: string
  deathday: string | null
  image: Image
  gender: string
}

export type Country = {
  name: string
  code: string
  timezone: string
}

export type Character = {
  id: number
  url: string
  name: string
  image: Image
}

export type SlideItem = Pick<Movie, 'id' | 'name'> & { rating: number; image: string }

export type CastItemProps = Pick<Person, 'name'> & { image: string }
