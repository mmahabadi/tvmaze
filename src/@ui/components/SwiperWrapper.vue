<script setup lang="ts">
import { Swiper } from 'swiper/vue'
import { Autoplay, Pagination, Navigation } from 'swiper/modules'
import { type SwiperOptions } from '@/@ui/types'
import { defineProps, computed } from 'vue'

import 'swiper/css'
import 'swiper/css/autoplay'
import 'swiper/css/navigation'
import 'swiper/css/pagination'
import 'swiper/css/scrollbar'

const props = defineProps<Partial<SwiperOptions>>()

const swiperStyle = {
  '--swiper-navigation-color': 'rgb(20 184 166 / var(--tw-text-opacity))',
  '--swiper-pagination-color': 'rgb(20 184 166 / var(--tw-text-opacity))'
}
const swiperOptions = computed(() => ({
  ...props
}))

const convertToInt = (prop: number | undefined) => {
  return +(prop ?? 0)
}

const breakpoints = computed(() => ({
  0: {
    slidesPerView: Math.min(1, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  768: {
    slidesPerView: Math.min(2, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  1024: {
    slidesPerView: Math.min(3, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(20, convertToInt(props.spaceBetween))
  },
  1280: {
    slidesPerView: Math.min(5, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(30, convertToInt(props.spaceBetween))
  },
  1536: {
    slidesPerView: convertToInt(props.slidesPerView),
    spaceBetween: convertToInt(props.spaceBetween)
  }
}))
</script>
<template>
  <swiper
    :style="swiperStyle"
    :modules="[Autoplay, Pagination, Navigation]"
    v-bind="swiperOptions"
    :breakpoints="breakpoints"
  >
    <slot></slot>
  </swiper>
</template>
