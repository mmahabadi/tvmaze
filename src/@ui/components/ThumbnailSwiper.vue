<script setup lang="ts">
import { SwiperSlide, Swiper } from 'swiper/vue'
import { Autoplay, Pagination, Navigation } from 'swiper/modules'
import { ref } from 'vue'
import ThumbnailItem from './ThumbnailItem.vue'
import { SlideItem } from '../types'

import 'swiper/css'
import 'swiper/css/autoplay'
import 'swiper/css/navigation'
import 'swiper/css/pagination'
import 'swiper/css/scrollbar'

defineProps<{ slides: SlideItem[] }>()

const modules = ref([Autoplay, Pagination, Navigation])
</script>
<template>
  <swiper
    :centeredSlides="true"
    :slides-per-view="5"
    :space-between="50"
    :navigation="true"
    :loop="true"
    :modules="modules"
    :scrollbar="{ draggable: true }"
  >
    <!-- :autoplay="{
      delay: 2500,
      disableOnInteraction: false
    }" -->
    <swiper-slide v-for="slide in slides" :key="slide.id">
      <div>
        <ThumbnailItem :item="slide" />
      </div>
    </swiper-slide>
  </swiper>
</template>
