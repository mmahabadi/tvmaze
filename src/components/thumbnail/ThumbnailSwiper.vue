<script setup lang="ts">
import { SwiperSlide } from 'swiper/vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'
import ThumbnailItem from './ThumbnailItem.vue'
import { SlideItem } from '@/types'
import { SwiperOptions } from '@/@ui/types'

defineProps<{ slides: SlideItem[] }>()

const options = {
  slidesPerView: 4,
  spaceBetween: 50,
  navigation: true,
  loop: true
} as SwiperOptions
</script>
<template>
  <SwiperWrapper v-bind="options">
    <swiper-slide v-for="slide in slides" :key="slide.id">
      <ThumbnailItem :item="slide" />
    </swiper-slide>
  </SwiperWrapper>
</template>
