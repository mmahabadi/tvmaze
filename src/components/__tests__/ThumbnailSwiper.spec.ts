import { describe, it, expect, beforeEach } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { SwiperSlide } from 'swiper/vue'
import ThumbnailSwiper from '../thumbnail/ThumbnailSwiper.vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'
import { type SlideItem } from '@/types'

describe('ThumbnailSwiper', () => {
  let wrapper: VueWrapper<unknown>
  const props = {
    loading: false,
    slidesPerView: 4,
    slides: [
      {
        id: 1,
        name: 'Inception',
        image: 'Inception.jpg',
        rating: 8.5
      }
    ] as SlideItem[]
  }

  beforeEach(() => {
    wrapper = mount(ThumbnailSwiper, {
      props,
      global: {
        stubs: {
          SwiperSlide,
          SwiperWrapper,
          ThumbnailItem: true
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
