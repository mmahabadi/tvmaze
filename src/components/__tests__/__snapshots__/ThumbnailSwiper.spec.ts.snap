// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThumbnailSwiper > renders correctly 1`] = `
"<div>
  <div class="swiper swiper-initialized swiper-horizontal" style="--swiper-navigation-color: rgb(20 184 166 / var(--tw-text-opacity)); --swiper-pagination-color: rgb(20 184 166 / var(--tw-text-opacity));" lazy="false">
    <div class="swiper-wrapper" style="transition-duration: 0ms; transition-delay: 0ms; transform: translate3d(NaNpx, 0px, 0px);">
      <div class="swiper-slide">
        <thumbnail-item-stub item="[object Object]"></thumbnail-item-stub>
        <!---->
      </div>
    </div>
    <div class="swiper-button-prev"></div>
    <div class="swiper-button-next"></div>
    <!---->
    <!---->
  </div>
</div>"
`;
